### PR TITLE
runtime: allow direct-assigned volumes when block device use is disabled

### PR DESCRIPTION
**Pull request by antonipp**
# Description

I have a use-case where I would like to generally keep `disable_block_device_use = true` by default for security reasons (described in https://github.com/kata-containers/kata-containers/security/advisories/GHSA-5fc8-gg7w-3g5c +  https://github.com/kata-containers/kata-containers/commit/20ca4d2) but I would still like to be able to have direct-assigned volumes plugged as block devices and not use virtiofs for performance reasons.

Direct-assigned volumes should be safe to hotplug because the CSI driver writes an explicit volume -> device mapping in `mountInfo.json` so there's no risk of inadvertently exposing host data (that's what the commit was trying to protect from)

This PR makes Kata read `mountInfo.json` first and if an explicit direct-assigned volume is present, it's honored even when `disable_block_device_use = true`. Everything else is unchanged: the rootfs guessing stays disabled, and ordinary bind mounts still get no block device. I got an AI agent write a bunch of unit tests to confirm that it's working as intended (so most of this PR is tests essentially)

btw I saw that someone else tried to implement a similar functionality in https://github.com/kata-containers/kata-containers/pull/12863 but my change is way more minimal, because here I'm _only_ allowing direct-assigned volumes and everything else stays blocked. IIUC there was some concern in that PR review because the logic was relying on the `IsFileBlockDevice` annotation for safety decisions but that shouldn't happen here because we are only relying on `mountInfo.json`

### CODE REVIEW

**Review by sprt — COMMENTED** [review]
Thanks @antonipp! Can you check if the same fix is needed in runtime-rs and implement it there as well if so?

**Comment by sprt** [comment]
BTW, I'd like to see a human-written regression analysis that describes why this change is NOT a regression of the GHSA you linked (before/after bullet points are easier to read than prose).

Plus there was a long discussion in https://github.com/kata-containers/kata-containers/pull/12863 so it'd be nice to briefly explain
* the difference in approach in this PR,
* why the concerns don't apply to this PR.

Please don't use AI here as that's generally hard to read.

**Comment by antonipp** [comment]
LOL I actually wrote the PR description myself 😅 Maybe I start sounding like an AI...

But to sum it up, the GHSA I linked to is a rootfs vulnerability. My PR does not change the behavior for rootfs mounts at all so when `disable_block_device_use` is set to `true`, there is no way for a rootfs to be mounted as a block device. 

So, the current behaviour on `main` with `disable_block_device_use = true`:
- rootfs block device hotplug is denied 
- direct-assigned volume block device hotplug is denied
- emptyDir block device hotplug is still allowed as an exception (I actually traced the git history and that's an exception you added in https://github.com/kata-containers/kata-containers/commit/a4fd32a29#diff-c7fd8974bc2a8b280571b84ffe785d6cbcd7b4a9d0217527ef1fe5c2c060f3c7)

After my PR:
- rootfs block device hotplug is still denied unconditionally 
- direct-assigned volume block device hotplug is now allowed as an exception
- emptyDirs are still allowed as an exception as well

It's the logic I put in `allowBlockDeviceForMount()`.

When it comes to https://github.com/kata-containers/kata-containers/pull/12863, the approach that the author took is broader than mine, i.e. "gating non-rootfs devices on hypervisor capabilities only". And the main problem was this heuristic: https://github.com/kata-containers/kata-containers/pull/12863#discussion_r3102581877 . So the code was making a safety decision based on the `IsFileBlockDevice` annotation and it could be bypassed by a rootfs without that annotation. I don't care about that all, so I am still blocking all rootfs block device mounts regardless and I only allow direct-assigned volumes with an explicit `mountInfo.json`.

Sorry if I'm being too verbose again 🤷 

And thanks I completely forgot about the Rust version, we don't use it currently so I focused on Go but I'll have a look.

**Comment by sprt** [comment]
Thanks this is a wonderful explanation and it would be great to include it in the commit message, along with a runtime-rs implementation!

I need to think more about the security implications but I'm very supportive of this change and I'd like to merge this as soon as reasonable.

**Comment by antonipp** [comment]
Ok cool, thanks. Updated the commit message with extra context.

I also had a look at the Rust runtime and I think this change is not really applicable there. AFAIU, the shape of the code is different over there, the flag is defined here (and it specifically mentions rootfs btw):
https://github.com/kata-containers/kata-containers/blob/97c7c9be6892cc973dc3e75c0a042015032163df/src/libs/kata-types/src/config/hypervisor/mod.rs#L230-L237

but it looks like direct-assigned devices are treated completely separately from this flag over here:
https://github.com/kata-containers/kata-containers/blob/97c7c9be6892cc973dc3e75c0a042015032163df/src/runtime-rs/crates/resource/src/volume/mod.rs#L138-L146 (at least I don't see this flag being involved in that path)
I haven't tried it out on a live machine but from what I can see, I think that the logic that I'm adding in the current PR for the Go runtime should _already_ work for the Rust runtime (so could be an incentive for us to try it out actually)

rootfs itself is handled in a different function here: https://github.com/kata-containers/kata-containers/blob/97c7c9be6892cc973dc3e75c0a042015032163df/src/runtime-rs/crates/resource/src/rootfs/mod.rs#L131-L139

and the `is_block_rootfs` check is here: https://github.com/kata-containers/kata-containers/blob/97c7c9be6892cc973dc3e75c0a042015032163df/src/runtime-rs/crates/resource/src/rootfs/block_rootfs.rs#L145

I'm not that familiar with the codebase but it does look like the original GHSA only applied to the Go runtime, the logic for rootfs handling seems different in the two runtimes

File: src/runtime/virtcontainers/container.go [hunk 1]
```diff
@@ -650,6 +650,22 @@ func (c *Container) createBlockDevices(ctx context.Context) error {
 			continue
 		}
```
Comment (sprt) [inline src/runtime/virtcontainers/container.go:652]:
```suggestion
		// If block rootfs is disabled, we only allow block emptyDirs
		// and direct-assigned volumes. This doesn't regress 20ca4d2,
		// which only pertains to rootfses.
		isBlockEmptyDir := isBlockEmptyDirMode(c.sandbox.config.EmptyDirMode) && IsDiskEmptyDir(c.mounts[i].Source)
		if !c.checkBlockDeviceSupport(ctx) && !isBlockEmptyDir && mntInfo == nil {
			c.Logger().Warn("Block-based rootfs is not supported")
			continue
		}
```

**Review by sprt — COMMENTED** [review]
Thanks @antonipp! I now understand the functional regression in runtime-go and I agree that runtime-rs is not affected. Re: runtime-rs, PTAL at https://github.com/kata-containers/kata-containers/pull/13682 when you get a chance.

I also agree with the spirit of your change, but I'm uncomfortable merging the code as is since:

* You're modifying the semantics of `checkBlockDeviceSupport()` but it's used by other callers.
* I think the diff could be smaller: see my code suggestion below - I believe adding only that code is enough to fix the functional regression. But please keep me honest here 🙂

File: src/runtime/virtcontainers/container.go (hunk 1 above)
Comment (antonipp) [reply src/runtime/virtcontainers/container.go:652]: Ok I gave this a try locally, it _does_ solve my original problem, however it makes a couple of minor edge case tests I added fail:
```
--- FAIL: TestCreateBlockDevicesDirectVolume/direct_volume_without_agent_support
    Error: Should be empty, but was 43df5e11157b16a3
    Error: expected: ".../001/volume"
           actual  : ".../001/disk.img"
--- FAIL: TestCreateBlockDevicesDirectVolume/direct_volume_without_hypervisor_hotplug_support
    Error: Should be empty, but was 50a668f30fcc614b
```

It's essentially trying to hotplug a direct volume even if the hypervisor or the Kata agent don't support it (cause `mntInfo == nil` wins vs `!c.checkBlockDeviceSupport(ctx)` check). Actually that's why I extracted the capability check from `checkBlockDeviceSupport()`. And to your first point, I changed `checkBlockDeviceSupport()` from:
```
if !DisableBlockDeviceUse 
  if agentCaps && hypervisorCaps 
    return true 
return false
```

to 

```
return !DisableBlockDeviceUse && checkBlockDeviceSupport
```
where `checkBlockDeviceSupport()` logic is essentially `agentCaps && hypervisorCaps`.

I see that this function is not really tested though, so I don't mind reverting it to the previous state, but we will have some incomplete edge case coverage. Otherwise, I can also add some tests for the function.

Also this test failed:
```
--- FAIL: TestCreateBlockDevicesDirectVolume/direct_volume_with_an_empty_device
    Error: expected: ".../001/volume"
           actual  : ""
```
It's another edge case, this time because of a malformed `mountInfo.json`:
```json
{"volume-type":"block","device":"","fstype":"ext4"}
```
With this version it tries to hotplug this as a block device and fails, whereas in my version there is an extra check for `mntInfo.Device != ""`  so it skips the block device hotplug attempt and is still able to get the directory via virtiofs. I guess it's a very improbable edge case but still.

Comment (antonipp) [reply src/runtime/virtcontainers/container.go:652]: @sprt have you been able to take a look?
